components/booking/TimeOptions.tsx

```diff
@@ -1,48 +1,56 @@
 // TODO: replace headlessui with radix-ui
 import { Switch } from "@headlessui/react";
+import { useRouter } from "next/router";
 import { FC, useEffect, useState } from "react";
 import TimezoneSelect, { ITimezoneOption } from "react-timezone-select";
 
 import classNames from "@lib/classNames";
+import { timeZone } from "@lib/clock";
 import { useLocale } from "@lib/hooks/useLocale";
-
-import { is24h, timeZone } from "../../lib/clock";
+import { useToggleQuery } from "@lib/hooks/useToggleQuery";
 
 type Props = {
   localeProp: string;
   onSelectTimeZone: (selectedTimeZone: string) => void;
-  onToggle24hClock: (is24hClock: boolean) => void;
 };
 
 const TimeOptions: FC<Props> = (props) => {
   const [selectedTimeZone, setSelectedTimeZone] = useState("");
   const [is24hClock, setIs24hClock] = useState(false);
   const { t } = useLocale({ localeProp: props.localeProp });
+  const hour12 = useToggleQuery("hour12", { zero: true });
+  const router = useRouter();
 
   useEffect(() => {
-    setIs24hClock(is24h());
+    setIs24hClock(
+      (router.query.hour12 !== undefined && !hour12.isOn) ||
+        !new Intl.DateTimeFormat(props.localeProp, { timeStyle: "short" })
+          .format(new Date())
+          .match(/\d+:\d+ (AM|PM)/)
+    );
```

Comment:
Can you explain this to me? @emrysal

Reply:
This detects the current "hour12" formatting - based on locale (Eg. "en") you end up with a string like "9:00 AM" but in "pl" you get "09:00". Unfortunately Intl does not detect these things so if you do something like `new Locale("en").hour12` you get back an absurdly helpful `undefined`. This all means there is no real way to figure out whether a user locale is AM/PM or 24H...

Except in this way, because this checks what the string is formatted as by Intl, so if !"9:00 AM" -> the user is in a locale that shows the time as non-AM/PM.